### ISSUE TITLE
clock skew error with grouped rules

### ISSUE THREAD

**Issue by mbhall88** [snakemake/snakemake#3165]
There is a rule `download` which downloads a FASTQ. The output of that rule is marked `temp()`. Then, there are three rules which can run in parallel, which all use this FASTQ. After those rules finish, the FASTQ is removed. However, I get this clock skew error ~25% of the time...

Importantly, I don't see this error if I run without the grouping

Here's the last bit of the log
```
5 of 5 steps (100%) done
Storing output in storage.
  + Exception Group Traceback (most recent call last):

  |   File "/home/mihall/sw/miniforge3/envs/gsize/lib/python3.12/site-packages/snakemake/cli.py", line 2100, in args_to_api
  |     dag_api.execute_workflow(

  |   File "/home/mihall/sw/miniforge3/envs/gsize/lib/python3.12/site-packages/snakemake/api.py", line 595, in execute_workflow
  |     workflow.execute(

  |   File "/home/mihall/sw/miniforge3/envs/gsize/lib/python3.12/site-packages/snakemake/workflow.py", line 1264, in execute
  |     raise e
  |   File "/home/mihall/sw/miniforge3/envs/gsize/lib/python3.12/site-packages/snakemake/workflow.py", line 1260, in execute
  |     success = self.scheduler.schedule()
  |               ^^^^^^^^^^^^^^^^^^^^^^^^^

  |   File "/home/mihall/sw/miniforge3/envs/gsize/lib/python3.12/site-packages/snakemake/scheduler.py", line 194, in schedule
  |     self._finish_jobs()

  |   File "/home/mihall/sw/miniforge3/envs/gsize/lib/python3.12/site-packages/snakemake/scheduler.py", line 401, in _finish_jobs
  |     async_run(postprocess())

  |   File "/home/mihall/sw/miniforge3/envs/gsize/lib/python3.12/site-packages/snakemake/common/__init__.py", line 89, in async_run
  |     return asyncio.run(coroutine)
  |            ^^^^^^^^^^^^^^^^^^^^^^

  |   File "/home/mihall/sw/miniforge3/envs/gsize/lib/python3.12/asyncio/runners.py", line 194, in run
  |     return runner.run(main)
  |            ^^^^^^^^^^^^^^^^

  |   File "/home/mihall/sw/miniforge3/envs/gsize/lib/python3.12/asyncio/runners.py", line 118, in run
  |     return self._loop.run_until_complete(task)
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  |   File "/home/mihall/sw/miniforge3/envs/gsize/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
  |     return future.result()
  |            ^^^^^^^^^^^^^^^

  |   File "/home/mihall/sw/miniforge3/envs/gsize/lib/python3.12/site-packages/snakemake/scheduler.py", line 363, in postprocess
  |     await job.postprocess(

  |   File "/home/mihall/sw/miniforge3/envs/gsize/lib/python3.12/site-packages/snakemake/jobs.py", line 1508, in postprocess
  |     async with asyncio.TaskGroup() as tg:

  |   File "/home/mihall/sw/miniforge3/envs/gsize/lib/python3.12/asyncio/taskgroups.py", line 145, in __aexit__
  |     raise me from None

  | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)

  +-+---------------- 1 ----------------

    | Traceback (most recent call last):

    |   File "/home/mihall/sw/miniforge3/envs/gsize/lib/python3.12/site-packages/snakemake/jobs.py", line 1148, in postprocess
    |     await self.dag.check_and_touch_output(

    |   File "/home/mihall/sw/miniforge3/envs/gsize/lib/python3.12/site-packages/snakemake/dag.py", line 730, in check_and_touch_output
    |     await self.check_output_mtime(job, expanded_output)

    |   File "/home/mihall/sw/miniforge3/envs/gsize/lib/python3.12/site-packages/snakemake/dag.py", line 752, in check_output_mtime
    |     raise WorkflowError(

    | snakemake_interface_common.exceptions.WorkflowError: Output /data/scratch/projects/punim2009/lrge/paper/results/estimates/genomescope/SRR/216/303/SRR21630359/SRR21630359.size has older modification time (2024-10-22 11:20:32.226016) than input /data/scratch/projects/punim2009/lrge/paper/results/downloads/SRR/216/303/SRR21630359/SRR21630359.fastq.gz (2024-10-22 11:20:32.969082). This could indicate a clock skew problem in your network and would trigger a rerun of this job in the next execution and should therefore be fixed on system level. System time: 2024-10-22 11:20:32.975679 (rule estimate_genomescope, line 156, /data/scratch/projects/punim2009/lrge/paper/workflow/rules/estimate.smk)
```

_Originally posted by @mbhall88 in https://github.com/snakemake/snakemake/issues/3097#issuecomment-2428018467_

**Comment by mbhall88** [comment]
So doing some more digging today, the problem seems to be that when grouped, there is a DAG execution of the `check_and_touch_output` function https://github.com/snakemake/snakemake/blob/cd6d035cf4e3ddcc671abc9c0517fafe6f2727ee/snakemake/dag.py#L671

in which the list of output files https://github.com/snakemake/snakemake/blob/cd6d035cf4e3ddcc671abc9c0517fafe6f2727ee/snakemake/dag.py#L681

includes files that are inputs for some rules also. In my example from above, the download rule, which outputs a fastq tagged as temp is an input to the estimate rules, after which time the fastq is used by no other rules.

So what ends up happening is the fastq file gets touched, along with all the other output files, here https://github.com/snakemake/snakemake/blob/cd6d035cf4e3ddcc671abc9c0517fafe6f2727ee/snakemake/dag.py#L727

This then, in some cases, causes the fastq, which is an input for the estimate rules, to have a newer mtime than the estimate output files.

I guess we maybe need some logic in the touch for loop that excludes touching files which are inputs to generate other output files in that list. I'm not sure what the best way to approach this is though - as in whether we have access to this information at the time of touching the files

**Comment by johanneskoester** [comment]
That makes sense! Thanks a lot for the detective work!

**Comment by johanneskoester** [comment]
This should fix it: https://github.com/snakemake/snakemake/pull/3181

**Comment by freekvh** [comment]
@johanneskoester Hi, I'm not sure if this is the right place, but I have the same issue, and I am still experiencing it in 8.25.2, Maybe I'm reading it wrong but that version was merged to main on nov 1 and v8.25.2 was tagged on nov 5...

```bash
$ conda list | grep snake
snakemake                 8.25.2               hdfd78af_0    bioconda
snakemake-executor-plugin-slurm 0.11.2             pyhdfd78af_0    bioconda
snakemake-executor-plugin-slurm-jobstep 0.2.1              pyhdfd78af_0    bioconda
snakemake-interface-common 1.17.4             pyhdfd78af_0    bioconda
snakemake-interface-executor-plugins 9.3.2              pyhdfd78af_0    bioconda
snakemake-interface-report-plugins 1.1.0              pyhdfd78af_0    bioconda
snakemake-interface-storage-plugins 3.3.0              pyhdfd78af_0    bioconda
snakemake-minimal         8.25.2             pyhdfd78af_0    bioconda
```

The (end of) the log:

```bash
[Fri Nov  8 22:04:13 2024]
Finished job 0.
2 of 2 steps (100%) done
Storing output in storage.
  + Exception Group Traceback (most recent call last):

  |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/site-packages/snakemake/cli.py", line 2158, in args_to_api
  |     dag_api.execute_workflow(

  |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/site-packages/snakemake/api.py", line 595, in execute_workflow
  |     workflow.execute(

  |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/site-packages/snakemake/workflow.py", line 1264, in execute
  |     raise e

  |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/site-packages/snakemake/workflow.py", line 1260, in execute
  |     success = self.scheduler.schedule()
  |               ^^^^^^^^^^^^^^^^^^^^^^^^^

  |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/site-packages/snakemake/scheduler.py", line 194, in schedule
  |     self._finish_jobs()

  |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/site-packages/snakemake/scheduler.py", line 401, in _finish_jobs
  |     async_run(postprocess())

  |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/site-packages/snakemake/common/__init__.py", line 90, in async_run
  |     return asyncio.run(coroutine)
  |            ^^^^^^^^^^^^^^^^^^^^^^

  |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/asyncio/runners.py", line 194, in run
  |     return runner.run(main)
  |            ^^^^^^^^^^^^^^^^

  |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/asyncio/runners.py", line 118, in run
  |     return self._loop.run_until_complete(task)
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
  |     return future.result()
  |            ^^^^^^^^^^^^^^^

  |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/site-packages/snakemake/scheduler.py", line 363, in postprocess
  |     await job.postprocess(

  |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/site-packages/snakemake/jobs.py", line 1518, in postprocess
  |     async with asyncio.TaskGroup() as tg:
  |                ^^^^^^^^^^^^^^^^^^^

  |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/asyncio/taskgroups.py", line 145, in __aexit__
  |     raise me from None

  | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)

  +-+---------------- 1 ----------------

    | Traceback (most recent call last):

    |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/site-packages/snakemake/jobs.py", line 1187, in postprocess
    |     raise e

    |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/site-packages/snakemake/jobs.py", line 1151, in postprocess
    |     await self.dag.check_and_touch_output(

    |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/site-packages/snakemake/dag.py", line 737, in check_and_touch_output
    |     await self.check_output_mtime(job, expanded_output)

    |   File "/home/fvhemert/miniforge3/envs/snaqs/lib/python3.12/site-packages/snakemake/dag.py", line 759, in check_output_mtime
    |     raise WorkflowError(

    | snakemake_interface_common.exceptions.WorkflowError: Output analyzed/0053_P2017BB3S20R_S2.genes.tpm4.results has older modification time (2024-11-08 22:04:13.137841) than input analyzed/salmon_0053_P2017BB3S20R_S2/quant.genes.sf (2024-11-08 22:04:13.261841). This could indicate a clock skew problem in your network and would trigger a rerun of this job in the next execution and should therefore be fixed on system level. System time: 2024-11-08 22:04:13.264866 (rule tpm4_normalization, line 1, /gpfs/scratch1/shared/fvhemert/temp/tests2/rules/tpm4_normalization_salmon.smk)

    +------------------------------------

```

**Comment by freekvh** [comment]
I just tested with 8.25.3 and it works for me now. I probably misunderstand how this fix moved to a release. Apologies.

**Comment by freekvh** [comment]
> I just tested with 8.25.3 and it works for me now. I probably misunderstand how the fix moved to a release. Apologies.

Ok, the bug is not fixed, it was a lucky hit I guess, as mentioned in the original issue, it does not happen every run...

```bash
snakemake_interface_common.exceptions.WorkflowError: Output analyzed/0053_P2017BB3S20R_S2.genes.tpm4.results has older modification time (2024-11-12 09:13:48.685712) than input analyzed/salmon_0053_P2017BB3S20R_S2/quant.genes.sf (2024-11-12 09:13:48.811711). This could indicate a clock skew problem in your network and would trigger a rerun of this job in the next execution and should therefore be fixed on system level. System time: 2024-11-12 09:13:48.816885 (rule tpm4_normalization, line 1, /gpfs/scratch1/shared/fvhemert/temp/tests/rules/tpm4_normalization_salmon.smk)
```

**Comment by obaeza16** [comment]
Hello, I was raising the same error in my snakemake run and I detected it was due to empty output declarations in a rule. Perhaps it it happening to you too? @freekvh 

And this was with the latest, 8.25.3 Snakemake version:
```
conda list | grep snake
snakemake                 8.25.3               hdfd78af_0    bioconda
snakemake-interface-common 1.17.4             pyhdfd78af_0    bioconda
snakemake-interface-executor-plugins 9.3.2              pyhdfd78af_0    bioconda
snakemake-interface-report-plugins 1.1.0              pyhdfd78af_0    bioconda
snakemake-interface-storage-plugins 3.3.0              pyhdfd78af_0    bioconda
snakemake-minimal         8.25.3             pyhdfd78af_0    bioconda
```

**Comment by freekvh** [comment]
> Hello, I was raising the same error in my snakemake run and I detected it was due to empty output declarations in a rule. Perhaps it it happening to you too? @freekvh

Nope, my output declarations are not empty, these are the grouped rules:

```python
rule salmon:
    input:
        read1 = rules.trim_galore.output.fastq_read1,
        read2 = rules.trim_galore.output.fastq_read2,
    output:
        genes_quantification = os.path.join(config['Directories']['analyzed_dir'], 'salmon_{sample}', 'quant.genes.sf'),
    threads: 12
    conda:
        "../envs/salmon.yaml"
    params:
        salmon_index = os.path.join(config['Global_directories']['reference_files'], config['References']['salmon_index']),
        gtf_file = os.path.join(config['Global_directories']['reference_files'], config['References']['gtf_file']),
    group:
        "quantification"    
    shell:
        '''
        salmon quant \
        --index {params.salmon_index} \
        --geneMap {params.gtf_file} \
        --libType A \
        --mates1 {input.read1} \
        --mates2 {input.read2} \
        --validateMappings \
        --threads {threads} \
        --output {config[Directories][analyzed_dir]}/salmon_{wildcards.sample}
        '''
```

```python
rule tpm4_normalization:
    input:
        genes_quantification = rules.salmon.output.genes_quantification
    output:
        tpm4_genes_results = os.path.join(
            config['Directories']['analyzed_dir'], 
            '{sample}.genes.tpm4.results'
        )
    group:
        "quantification"
    run:
        Some python code here.
```

**Comment by freekvh** [comment]
@obaeza16 Btw, do you know, should I open a new issue, or just be more patient and wait for a response here? I'm not sure of the etiquette here...

**Comment by freekvh** [comment]
Thanx.

**Comment by mbhall88** [comment]
@johanneskoester I can confirm this issue is still present despite the changes in https://github.com/snakemake/snakemake/pull/3181

**Comment by johanneskoester** [comment]
On it, I think I know where the problem is.

**Comment by johanneskoester** [comment]
This PR might fix it: https://github.com/snakemake/snakemake/pull/3220

**Comment by johanneskoester** [comment]
Could you try it out and let me know whether it helps?

**Comment by freekvh** [comment]
@johanneskoester I have run my tests pipelines 3 times now without issue using this latest PR.

I'll now deploy the pipeline with some more/larger datasets but I think it is fixed (it was a stochastic error but am pretty confident by now).

**Comment by freekvh** [comment]
Ok, I have run a lot of pipelines now and haven't seen the error anymore, as far as I'm concerned this is fixed.
